### PR TITLE
code returned by fallback.ServeDNS is not correct one

### PR DESCRIPTION
**Pull request by mparamonau**

### CODE REVIEW

**Review by yuewko — COMMENTED** [review]
LGTM!

File: fallback.go [hunk 1]
```diff
@@ -43,22 +43,27 @@ func New(trace plugin.Handler) (f *Fallback) {
...
+	if nw.Msg != nil {
+		rules_index = nw.Msg.Rcode
 	}
-	if u, ok := f.rules[rcode]; ok {
+
+	if u, ok := f.rules[rules_index]; ok {
 		p := f.proxy.New(f.trace, u)
 		if p == nil {
-			return rcode, fmt.Errorf("cannot create fallback proxy")
+			return rules_index, fmt.Errorf("cannot create fallback proxy")
```
Comment (fturib) [inline fallback.go]: I guess if we throw an error, we should return SERVFAIL .. no ? Also because initial writer w is not filled.

File: fallback.go [hunk 2]
```diff
@@ -43,22 +43,27 @@ func New(trace plugin.Handler) (f *Fallback) {
...
 		}
 		return p.ServeDNS(ctx, w, r)
 	}
-	w.WriteMsg(nw.Msg)
-	return rcode, nil
+	if nw.Msg != nil {
```
Comment (fturib) [inline fallback.go:63]: test nw.Msg != nil is for safe.
but the logic test would be err != nil.

=> if you return an Error, then you return (SERVAIL, err) and the MSG is not written .. - I am not sure here.

File: fallback.go (hunk 1 above)
Comment (mparamonau) [reply fallback.go]: I guess - yes, and this is what we do, in case of SERVFAIL the rcode is SERVFAIL as well as rules_index, isn't it?

File: fallback.go (hunk 2 above)
Comment (mparamonau) [reply fallback.go:63]: it seems it's not for safe: when we get SERVFAIL the nw.Msg is nil and if we don't have  appropriate rules, we mustn't do w.WriteMsg(nw.Msg) because of we will get SIGSEGV.

Comment (mparamonau) [reply fallback.go:63]: For example, in config we have next:
fallback NXDOMAIN . 10.6.0.6 {
      policy first
      }

but we get SERVFAIL, in this case nw.Msg.

File: fallback.go [hunk 3]
```diff
@@ -43,22 +43,27 @@ func New(trace plugin.Handler) (f *Fallback) {
 func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
 	nw := nonwriter.New(w)
 	rcode, err := plugin.NextOrFailure(f.Name(), f.Next, ctx, nw, r)
-	if err != nil {
```
Comment (fturib) [inline fallback.go:46]: do we change the semantic here ?
I mean : if there is an internal error (by one of the following handlers), then the reply will be : SERVFAIL, err

I think, before this code change, we were NOT taking into account a SERVFAIL which means a local error. 
Now, we are processing this error in "fallback". Is it expected ?

Comment (fturib) [reply fallback.go:46]: Ok, after reading the README.md of fallback, I think we want to handle this use case in Fallback.
So, this change of code is ok for me.

File: fallback.go (hunk 2 above)
Comment (mparamonau) [reply fallback.go:63]: And yea, instead of nw.Msg != nil we can check err == nil, do I need to modify it?

File: fallback.go (hunk 3 above)
Comment (fturib) [reply fallback.go:46]: Although here there is maybe a mismatch between:
- internal fatal errors returned by plugins (with no DNS semantic here)
- a SERVAIL DNS result, that would be built by a plugin of CoreDNS 

for instance, PROXY and FORWARD return SERVAIL + raise an error to indicates it cannot serve the msg.

Second: I am not sure what is expected for FALLBACK. 
Handle ALL SERVAIL that will be replied to caller ? or only the ones that have a DNS meaning ?

File: fallback.go (hunk 1 above)
Comment (fturib) [reply fallback.go]: hum.. no. you can have a rules_index that is REFUSE, you catch if in your fallback rule, call the f.proxy and get a NIL.

=> in that case, I would expect you to return SERVFAIL, "cannot create fallback", instead of REFUSE, "cannot create fallback".

**Comment by fturib** [comment]
fixing #4

File: fallback.go (hunk 1 above)
Comment (mparamonau) [reply fallback.go]: Oh, got it, you meant we need to return 

return dns.RcodeServerFailure, fmt.Errorf("cannot create fallback proxy")?

File: fallback.go (hunk 2 above)
Comment (fturib) [reply fallback.go:63]: I think would be good to check : err == nil && nw.Msg != nil

File: fallback.go (hunk 1 above)
Comment (fturib) [reply fallback.go]: yes.

File: fallback.go (hunk 2 above)
Comment (mparamonau) [reply fallback.go:63]: ok, sure

Comment (yuewko) [reply fallback.go:63]: I thought nw.Msg==nil is a legitimate message.

File: fallback.go (hunk 3 above)
Comment (yuewko) [reply fallback.go:46]: IMHO...

In general, `err` should be treated as fatal error. Therefore, I believe that the original check is legitimate, "EXCEPT":

As the description of the fallback plugin indicates, its purpose is "...to allow a fallback when, for example, the desired upstreams became unavailable"

Unfortunately, for the proxy plugin, there are at least a couple of cases where `err` is used in at least a couple of cases to indicate that "upstreams became unavailable":

- https://github.com/coredns/coredns/blob/master/plugin/proxy/proxy.go#L80
- https://github.com/coredns/coredns/blob/master/plugin/proxy/proxy.go#L157

Maybe we should treat those as special cases?

Comment (fturib) [reply fallback.go:46]: Looking into the definition of Plugin and parameters it should returned : plugin/plugin.go
I think:

1- the "Err" that maybe returned is only for logging.
2- the plugin has to reply a rcode that inform if the the reply MSG was written in the Response or not : SERVAIL/REFUSE/FORMEER/NOTIMP are taken for : "the reply was not written, the top level, will have to forge a reply message"
3- any other rcode means the reply MSG is written and it bring the final reply to the CLIENT (including rcode).

From there, IMO, there is no "fatal error" - only DNS rcode we want to handle with fallback. 
if the plan is to handle SERVFAIL, then we should handle it whatever reason raised the reply SERVFAIL.

In other words, I would just keep the new code proposed. and do not test at all the "if Err != nil".

File: fallback.go (hunk 2 above)
Comment (fturib) [reply fallback.go:63]: After reflexion, I guess you are right. if Msg != nil should be ok. (Err is only an information).

File: fallback.go (hunk 1 above)
Comment (mparamonau) [reply fallback.go]: fixed

**Review by fturib — APPROVED** [review]
After a long conversation ... 
and a small update ..
I am ok with these changes. 👍

**Comment by fturib** [comment]
@johnbelamaric : Could you merge this PR ?
